lightning/src/routing/scoring.rs

```diff
@@ -460,25 +458,466 @@ impl<T: Time> Readable for ChannelFailure<T> {
+		let (min_liquidity_offset_msat, max_liquidity_offset_msat) = if source < target {
+			(&self.min_liquidity_offset_msat, &self.max_liquidity_offset_msat)
+		} else {
+			(&self.max_liquidity_offset_msat, &self.min_liquidity_offset_msat)
+		};
+
+		DirectedChannelLiquidity {
+			min_liquidity_offset_msat,
+			max_liquidity_offset_msat,
+			capacity_msat,
+			last_updated: &self.last_updated,
+			now: T::now(),
+			half_life,
+		}
+	}
+
+	/// Returns a mutable view of the channel liquidity directed from `source` to `target` assuming
+	/// `capacity_msat`.
+	fn as_directed_mut(
+		&mut self, source: &NodeId, target: &NodeId, capacity_msat: u64, half_life: Duration
+	) -> DirectedChannelLiquidity<&mut u64, T, &mut T> {
+		let (min_liquidity_offset_msat, max_liquidity_offset_msat) = if source < target {
+			(&mut self.min_liquidity_offset_msat, &mut self.max_liquidity_offset_msat)
+		} else {
+			(&mut self.max_liquidity_offset_msat, &mut self.min_liquidity_offset_msat)
+		};
+
+		DirectedChannelLiquidity {
+			min_liquidity_offset_msat,
+			max_liquidity_offset_msat,
+			capacity_msat,
+			last_updated: &mut self.last_updated,
+			now: T::now(),
+			half_life,
+		}
+	}
+}
+
+impl<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> DirectedChannelLiquidity<L, T, U> {
+	/// Returns the success probability of routing the given HTLC `amount_msat` through the channel
+	/// in this direction.
+	fn success_probability(&self, amount_msat: u64) -> Probability {
+		let max_liquidity_msat = self.max_liquidity_msat();
+		let min_liquidity_msat = core::cmp::min(self.min_liquidity_msat(), max_liquidity_msat);
+		if amount_msat > max_liquidity_msat {
+			Probability::Zero
+		} else if amount_msat < min_liquidity_msat {
+			Probability::One
+		} else {
+			let numerator = max_liquidity_msat + 1 - amount_msat;
+			let denominator = max_liquidity_msat + 1 - min_liquidity_msat;
+			if numerator == denominator {
+				Probability::One
+			} else {
+				Probability::Ratio { numerator, denominator }
+			}
+		}
+	}
+
+	/// Returns the lower bound of the channel liquidity balance in this direction.
+	fn min_liquidity_msat(&self) -> u64 {
+		self.decayed_offset_msat(*self.min_liquidity_offset_msat)
+	}
+
+	/// Returns the upper bound of the channel liquidity balance in this direction.
+	fn max_liquidity_msat(&self) -> u64 {
+		self.capacity_msat
+			.checked_sub(self.decayed_offset_msat(*self.max_liquidity_offset_msat))
+			.unwrap_or(0)
+	}
+
+	fn decayed_offset_msat(&self, offset_msat: u64) -> u64 {
+		self.now.duration_since(*self.last_updated).as_secs()
+			.checked_div(self.half_life.as_secs())
+			.and_then(|decays| offset_msat.checked_shr(decays as u32))
+			.unwrap_or(0)
+	}
+}
+
+impl<L: DerefMut<Target = u64>, T: Time, U: DerefMut<Target = T>> DirectedChannelLiquidity<L, T, U> {
+	/// Adjusts the channel liquidity balance bounds when failing to route `amount_msat`.
+	fn failed_at_channel(&mut self, amount_msat: u64) {
```

Comment:
I just have to say that I really love to see this and the following functions being implemented here and that you will track the uncertainty / information gain properly! I will be really curios how much this also improves payment delivery in practice in comparison to the recent updates that we have seen from c-lightning using only probabilistic scoring but forgetting most of the learnt information directly

Reply:
Hopefully Cash App switches and uses the new one and we can get some good data from that!